{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71687918-bf53-4895-b8ee-f84c3a73bbf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "pd.set_option('display.max_rows', 500)\n",
    "import altair as alt\n",
    "from ete3 import NCBITaxa\n",
    "ncbi = NCBITaxa()\n",
    "#ncbi.update_taxonomy_database()\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26eab5fd-b276-4d60-8f3d-97223cddbb76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(\"/Users/anton/Downloads/Supplementary Tables (VGP-Galaxy).xlsx\",\"ST2 - Assembly full list\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2682ddbb-859c-4037-bf4b-43da4b2e226b",
   "metadata": {},
   "source": [
    "# DANGER HERE\n",
    "This slicing below is very arbitrary and depends on the content of the spreadsheet. Check every time!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8a447080-aa82-4108-9e00-738ae8aa50f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# At present the table is not harmonized, so some processing is required\n",
    "df = df.iloc[:73,:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "befed351-654a-4c3c-9248-9a437053a414",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If a species only has one assembly, it will be called hap1\n",
    "df.loc[ ( df['Assembly version'] != 'hap1' ) & ( df['Assembly version'] != 'hap2' ), 'Assembly version'] = 'hap1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5b7ab78f-2f5a-4be7-a2a3-8982dad49654",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "for sp, taxid in ncbi.get_name_translator(df['NCBI Species'].unique()).items():\n",
    "    lineage = ncbi.get_lineage(taxid[0])\n",
    "    for item in lineage:\n",
    "        rank = ncbi.get_rank([item])\n",
    "        for key,value in rank.items():\n",
    "            if value == 'phylum':\n",
    "                df.loc[ df['NCBI Species'] == sp ,'phylum'] = ncbi.translate_to_names([key])[0]\n",
    "            elif value == 'class':\n",
    "                df.loc[ df['NCBI Species'] == sp ,'class_']  = ncbi.translate_to_names([key])[0]\n",
    "            elif value == 'order':\n",
    "                df.loc[ df['NCBI Species'] == sp ,'order']  = ncbi.translate_to_names([key])[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aceeb05-441d-4c48-abe0-e3a96879f527",
   "metadata": {},
   "source": [
    "# DANGER HERE! \n",
    "For some stupid reason there is no `class` rank for Reptilia, so these need to be set manually:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5a0289d5-1c45-49d8-b3d8-29438b300be9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.loc[df['class_'].isna(), 'class_'] = \"Reptilia\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8eca95e0-7a92-4599-9cf8-00237e8bc00e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ToLID\n",
      "Species\n",
      "NCBI Species\n",
      "Common name\n",
      "Assembly version\n",
      "Pipeline version\n",
      "Curated\n",
      "Total bp\n",
      "Genome size¹\n",
      "Sequencing depth\n",
      "Heterozygosity¹\n",
      "Repeat content¹\n",
      "Sex (metadata)\n",
      "Heterogametic\n",
      "Mitogenome?\n",
      "Assembler\n",
      "Bionano\n",
      "Draft assembly folder\n",
      "Unnamed: 18\n",
      "Curated assembly folder\n",
      "Notes\n",
      "# scaffolds\n",
      "Total scaffold length\n",
      "Average scaffold length\n",
      "Scaffold N50\n",
      "Scaffold auN\n",
      "Scaffold L50\n",
      "Scaffold NG50\n",
      "Scaffold auNG\n",
      "Scaffold LG50\n",
      "Largest scaffold\n",
      "Smallest scaffold\n",
      "# contigs\n",
      "Total contig length\n",
      "Average contig length\n",
      "Contig N50\n",
      "Contig auN\n",
      "Contig L50\n",
      "Contig NG50\n",
      "Contig auNG\n",
      "Contig LG50\n",
      "Largest contig\n",
      "Smallest contig\n",
      "# gaps in scaffolds\n",
      "Total gap length in scaffolds\n",
      "Average gap length in scaffolds\n",
      "Gap N50 in scaffolds\n",
      "Gap auN in scaffolds\n",
      "Gap L50 in scaffolds\n",
      "Largest gap in scaffolds\n",
      "Smallest gap in scaffolds\n",
      "Base composition (A:C:G:T)\n",
      "GC content %\n",
      "# soft-masked bases\n",
      "# segments\n",
      "Total segment length\n",
      "Average segment length\n",
      "# gaps\n",
      "# paths\n",
      "Merqury completeness (hifi kmers) of contigs in ONE HAP (if pri/alt, then post-purging)\n",
      "Merqury completeness (hifi kmers) of contigs in BOTH HAPS (if pri/alt, then post-purging)\n",
      "%age complete single BUSCO genes in contigs (if pri/alt, then post-purging)\n",
      "%age complete duplicated BUSCO genes in contigs (if pri/alt, then post-purging)\n",
      "%age complete BUSCO genes in contigs (if pri/alt, then post-purging)\n",
      "phylum\n",
      "class_\n",
      "order\n"
     ]
    }
   ],
   "source": [
    "for _ in df.columns: print(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9c9dd72d-6fa6-4c4e-8969-ee298fc51563",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amphibia\n",
      "Aves\n",
      "Actinopteri\n",
      "Mammalia\n",
      "Lepidosauria\n",
      "Reptilia\n",
      "Chondrichthyes\n"
     ]
    }
   ],
   "source": [
    "for _ in df['class_'].unique(): print(_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "217d3ef4-3641-4112-aba2-918a67edd259",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phylo rank for prper ordering of classes:\n",
    "class_rank = {\n",
    "    'Actinopteri':2,\n",
    "    'Amphibia':3,\n",
    "    'Aves':6,\n",
    "    'Chondrichthyes':1,\n",
    "    'Lepidosauria':5,\n",
    "    'Mammalia':7,\n",
    "    'Reptilia':4\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "94dd2940-be44-46d9-95e1-c53415e6ec99",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in class_rank:\n",
    "    df.loc[df['class_']==key, 'c_rank'] = class_rank[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "da35f735-c7d8-4409-8708-670127c425b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.sort_values(by=['c_rank','order','NCBI Species','Assembly version'],ignore_index = True).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5442c35e-d39e-4ae3-8d6d-029900f948c9",
   "metadata": {},
   "source": [
    "# DANGER HERE\n",
    "\n",
    "The following three `groupby`'s need to be converted to a function. They are doing slightkly different things = hence the repeatitive code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8517ff70-6eda-4de0-9093-f6897942e454",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_order_chart = df.groupby(\n",
    "        'order'\n",
    "    ).agg(\n",
    "        mn = pd.NamedAgg(column='index',aggfunc ='min'),\n",
    "        mx = pd.NamedAgg(column='index',aggfunc = lambda x: x.max()+1),\n",
    "        mid = pd.NamedAgg(column='index',aggfunc = lambda x: (x.min()+((x.max()-x.min())/2))),\n",
    "        class_ = pd.NamedAgg(column='class_',aggfunc = max)\n",
    "    ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a133a15e-570c-42c2-8a2f-9276b4952c98",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>order</th>\n",
       "      <th>mn</th>\n",
       "      <th>mx</th>\n",
       "      <th>mid</th>\n",
       "      <th>class_</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Anura</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>18.5</td>\n",
       "      <td>Amphibia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Artiodactyla</td>\n",
       "      <td>53</td>\n",
       "      <td>60</td>\n",
       "      <td>56.0</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Atheriniformes</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4.0</td>\n",
       "      <td>Actinopteri</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Caprimulgiformes</td>\n",
       "      <td>33</td>\n",
       "      <td>34</td>\n",
       "      <td>33.0</td>\n",
       "      <td>Aves</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Carnivora</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "      <td>60.0</td>\n",
       "      <td>Mammalia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              order  mn  mx   mid       class_\n",
       "0             Anura  16  22  18.5     Amphibia\n",
       "1      Artiodactyla  53  60  56.0     Mammalia\n",
       "2    Atheriniformes   4   5   4.0  Actinopteri\n",
       "3  Caprimulgiformes  33  34  33.0         Aves\n",
       "4         Carnivora  60  61  60.0     Mammalia"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_order_chart.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7fd5e3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_class_chart = df.groupby(\n",
    "        'class_'\n",
    "    ).agg(\n",
    "        mn = pd.NamedAgg(column='index',aggfunc ='min'),\n",
    "        mx = pd.NamedAgg(column='index',aggfunc = lambda x: x.max()+1),\n",
    "        mid = pd.NamedAgg(column='index',aggfunc = lambda x: (x.min()+((x.max()-x.min())/2))),\n",
    "    ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "92159697-68cb-4a3c-9b54-bc6910417d0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class_</th>\n",
       "      <th>mn</th>\n",
       "      <th>mx</th>\n",
       "      <th>mid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Actinopteri</td>\n",
       "      <td>4</td>\n",
       "      <td>16</td>\n",
       "      <td>9.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Amphibia</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>18.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aves</td>\n",
       "      <td>33</td>\n",
       "      <td>53</td>\n",
       "      <td>42.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Chondrichthyes</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lepidosauria</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>29.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           class_  mn  mx   mid\n",
       "0     Actinopteri   4  16   9.5\n",
       "1        Amphibia  16  22  18.5\n",
       "2            Aves  33  53  42.5\n",
       "3  Chondrichthyes   0   4   1.5\n",
       "4    Lepidosauria  27  33  29.5"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_class_chart.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "508ac95d-4d95-4649-9bee-452e4ce2ee46",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_for_species_chart = df.groupby(\n",
    "        ['NCBI Species','Assembly version']\n",
    "    ).agg(\n",
    "        mn = pd.NamedAgg(column='index',aggfunc ='min'),\n",
    "        mx = pd.NamedAgg(column='index',aggfunc = lambda x: x.max()+1),\n",
    "        mid = pd.NamedAgg(column='index',aggfunc = lambda x: (x.min()+((x.max()-x.min())/2))),\n",
    "        species = pd.NamedAgg(column='NCBI Species',aggfunc = max),\n",
    "        class_ = pd.NamedAgg(column='class_',aggfunc = max),\n",
    "        order = pd.NamedAgg(column='order',aggfunc = max),\n",
    "        size = pd.NamedAgg(column='Genome size¹',aggfunc = lambda x: (x.max()/1000000)),\n",
    "        het = pd.NamedAgg(column='Heterozygosity¹',aggfunc = max),\n",
    "        rep = pd.NamedAgg(column='Repeat content¹',aggfunc = max),\n",
    "        s_ng50 = pd.NamedAgg(column='Scaffold NG50',aggfunc = lambda x: (x.max()/1000000)),\n",
    "        c_ng50 = pd.NamedAgg(column='Contig NG50',aggfunc = lambda x: (x.max()/1000000)),\n",
    "        sGap = pd.NamedAgg(column='Total gap length in scaffolds',aggfunc = lambda x: (x.max()/1000000)),\n",
    "    ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7b56e188-c89c-4ed8-b6fe-79143f621401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NCBI Species</th>\n",
       "      <th>Assembly version</th>\n",
       "      <th>mn</th>\n",
       "      <th>mx</th>\n",
       "      <th>mid</th>\n",
       "      <th>species</th>\n",
       "      <th>class_</th>\n",
       "      <th>order</th>\n",
       "      <th>size</th>\n",
       "      <th>het</th>\n",
       "      <th>rep</th>\n",
       "      <th>s_ng50</th>\n",
       "      <th>c_ng50</th>\n",
       "      <th>sGap</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Acridotheres tristis</td>\n",
       "      <td>hap1</td>\n",
       "      <td>39</td>\n",
       "      <td>40</td>\n",
       "      <td>39.0</td>\n",
       "      <td>Acridotheres tristis</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Passeriformes</td>\n",
       "      <td>1354.313347</td>\n",
       "      <td>0.558</td>\n",
       "      <td>30.0</td>\n",
       "      <td>29.697385</td>\n",
       "      <td>6.530360</td>\n",
       "      <td>3.114772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ammodramus caudacutus</td>\n",
       "      <td>hap1</td>\n",
       "      <td>40</td>\n",
       "      <td>41</td>\n",
       "      <td>40.0</td>\n",
       "      <td>Ammodramus caudacutus</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Passeriformes</td>\n",
       "      <td>1225.461744</td>\n",
       "      <td>0.236</td>\n",
       "      <td>21.4</td>\n",
       "      <td>39.973966</td>\n",
       "      <td>8.438390</td>\n",
       "      <td>4.089548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ammodramus nelsoni</td>\n",
       "      <td>hap1</td>\n",
       "      <td>41</td>\n",
       "      <td>42</td>\n",
       "      <td>41.0</td>\n",
       "      <td>Ammodramus nelsoni</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Passeriformes</td>\n",
       "      <td>1190.976425</td>\n",
       "      <td>0.571</td>\n",
       "      <td>21.3</td>\n",
       "      <td>40.159020</td>\n",
       "      <td>12.036358</td>\n",
       "      <td>5.142283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara ararauna</td>\n",
       "      <td>hap1</td>\n",
       "      <td>49</td>\n",
       "      <td>50</td>\n",
       "      <td>49.0</td>\n",
       "      <td>Ara ararauna</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Psittaciformes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara ararauna</td>\n",
       "      <td>hap2</td>\n",
       "      <td>50</td>\n",
       "      <td>51</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Ara ararauna</td>\n",
       "      <td>Aves</td>\n",
       "      <td>Psittaciformes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            NCBI Species Assembly version  mn  mx   mid  \\\n",
       "0   Acridotheres tristis             hap1  39  40  39.0   \n",
       "1  Ammodramus caudacutus             hap1  40  41  40.0   \n",
       "2     Ammodramus nelsoni             hap1  41  42  41.0   \n",
       "3           Ara ararauna             hap1  49  50  49.0   \n",
       "4           Ara ararauna             hap2  50  51  50.0   \n",
       "\n",
       "                 species class_           order         size    het   rep  \\\n",
       "0   Acridotheres tristis   Aves   Passeriformes  1354.313347  0.558  30.0   \n",
       "1  Ammodramus caudacutus   Aves   Passeriformes  1225.461744  0.236  21.4   \n",
       "2     Ammodramus nelsoni   Aves   Passeriformes  1190.976425  0.571  21.3   \n",
       "3           Ara ararauna   Aves  Psittaciformes          NaN    NaN   NaN   \n",
       "4           Ara ararauna   Aves  Psittaciformes          NaN    NaN   NaN   \n",
       "\n",
       "      s_ng50     c_ng50      sGap  \n",
       "0  29.697385   6.530360  3.114772  \n",
       "1  39.973966   8.438390  4.089548  \n",
       "2  40.159020  12.036358  5.142283  \n",
       "3        NaN        NaN       NaN  \n",
       "4        NaN        NaN       NaN  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_for_species_chart.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "84f7f4f4-0c08-4e36-89ec-f1eb15b72bcc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# This is super-ugly\n",
    "# It needs to be partitioned into separate functions\n",
    "\n",
    "domain_c = df['class_'].unique()\n",
    "range_c = ['#d73027','#fc8d59','#fee090','#ffffbf','#e0f3f8','#91bfdb','#4575b4']\n",
    "\n",
    "class_chart = alt.Chart(df_for_class_chart).mark_rect(opacity=1,stroke='black', strokeWidth=.5).encode(\n",
    "    y = alt.Y('mn:Q',scale=alt.Scale(domain=[0, 72],nice=False),axis=None),\n",
    "    y2 = 'mx:Q',\n",
    "    color=alt.Color('class_:N',scale=alt.Scale(domain=domain_c,range=range_c),legend=None),\n",
    ").properties(\n",
    "        height=900,\n",
    "        width=100,\n",
    "        title = 'Class'\n",
    ")\n",
    "\n",
    "class_text = class_chart.transform_calculate(\n",
    "    link='https://www.google.com/search?q=' + alt.datum.class_\n",
    ").mark_text(color='black',align='center',baseline=\"middle\",fontSize=12,fontWeight=\"bold\",dy=-5).encode(\n",
    "    y = 'mid:Q',\n",
    "    text='class_:N',\n",
    "    color=alt.value('black'),\n",
    "    href = 'link:N',\n",
    "    #color=alt.condition(\n",
    "    #    alt.datum['name']== \"Reptilia\",\n",
    "    #    alt.value(\"white\"),\n",
    "    #    alt.value(\"black\")\n",
    "    #)\n",
    ")\n",
    "\n",
    "order_chart = alt.Chart(df_for_order_chart).mark_rect(opacity=1,stroke='black', strokeWidth=.5).encode(\n",
    "    y = alt.Y('mn:Q',scale=alt.Scale(domain=[0, 72],nice=False),axis=None),\n",
    "    y2 = 'mx:Q',\n",
    "    color=alt.Color('class_:N',scale=alt.Scale(domain=domain_c,range=range_c),legend=None),\n",
    "    opacity=alt.Opacity('order:N',legend=None)\n",
    ").properties(\n",
    "        height=900,\n",
    "        width=100,\n",
    "    title = 'Order'\n",
    ")\n",
    "\n",
    "order_text = order_chart.transform_calculate(\n",
    "    link='https://www.google.com/search?q=' + alt.datum.order\n",
    ").mark_text(color='black',align='center',baseline=\"middle\",fontSize=10,dy=-5).encode(\n",
    "    y = 'mid:Q',\n",
    "    text='order:N',\n",
    "    color=alt.value('black'),\n",
    "    opacity=alt.value(1),\n",
    "    href = 'link:N',\n",
    "    #color=alt.condition(\n",
    "    #    alt.datum['name']== \"Reptilia\",\n",
    "    #    alt.value(\"white\"),\n",
    "    #    alt.value(\"black\")\n",
    "    #)\n",
    ")\n",
    "\n",
    "species_chart = alt.Chart(df_for_species_chart).mark_rect(opacity=1,stroke='black', strokeWidth=.5).encode(\n",
    "    y = alt.Y('mn:Q',scale=alt.Scale(domain=[0, 72],nice=False),axis=None),\n",
    "    y2 = 'mx:Q',\n",
    "    #x = alt.value(10)\n",
    "    color=alt.Color('class_:N',scale=alt.Scale(domain=domain_c,range=range_c),legend=None),\n",
    "    opacity=alt.Opacity('order:N',legend=None)\n",
    ").properties(\n",
    "        height=900,\n",
    "        width=350,\n",
    "    title='Species'\n",
    ")\n",
    "\n",
    "species_text = species_chart.transform_calculate(\n",
    "    link='https://www.google.com/search?q=' + alt.datum.species\n",
    ").mark_text(color='black',align='center',baseline=\"middle\",fontSize=8,dy=-5,fontStyle=\"italic\").encode(\n",
    "    y = 'mid:Q',\n",
    "    text='species:N',\n",
    "    href = 'link:N',\n",
    "    #color=alt.value('black'),\n",
    "    opacity=alt.value(1),\n",
    "    color=alt.condition(\n",
    "        alt.datum['class_']== \"Mammalia\",\n",
    "        alt.value(\"black\"),\n",
    "        alt.value(\"black\")\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f67172aa-85b9-4ca4-aa30-1e6b7e98fedc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def hmt(df,col,y,y2,op,scheme,m_value,format_,wid,title,stitle):\n",
    "    chart = alt.Chart(df,width=wid).mark_rect().encode(\n",
    "        y = alt.Y(y,scale=alt.Scale(domain=[0, 72],nice=False),axis=None),\n",
    "        y2 = y2,\n",
    "    color=alt.Color(col,scale=alt.Scale(scheme=scheme),legend=None),\n",
    "    ).properties(\n",
    "        height=900,\n",
    "        #width=50,\n",
    "        title={ \"text\": title,\"subtitle\":stitle }\n",
    "               \n",
    "    )\n",
    "    \n",
    "    text = chart.mark_text(color='black',align='center',baseline=\"middle\",fontSize=8,dy=-5).encode(\n",
    "    y = y,\n",
    "    text=alt.Text(col,format=format_),\n",
    "    #color=alt.value('black'),\n",
    "    opacity=alt.value(1),\n",
    "    color=alt.condition(\n",
    "        alt.datum[col] > m_value,\n",
    "        alt.value(\"white\"),\n",
    "        alt.value(\"black\")\n",
    "    )\n",
    "    )\n",
    "    return(chart + text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4da00046-7dad-4392-88f5-605e2f097ce5",
   "metadata": {},
   "outputs": [],
   "source": [
    "size = hmt(df_for_species_chart,'size','mn:Q','mx:Q','order','goldred',4000,\",.0f\",60,'Size','(Mb)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "471303a1-ac24-4047-859e-ce49fa1b02d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "het = hmt(df_for_species_chart,'het','mn:Q','mx:Q','order','yellowgreen',.5,\",.2f\",60,'Het','(%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "cfa831d1-b28d-45b1-b522-78bd314d373a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rep = hmt(df_for_species_chart,'rep','mn:Q','mx:Q','order','yellowgreen',50,\",.1f\",60,'Repeat','(%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb400287-3f2f-4297-9a76-a356311fb640",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_ng50 = hmt(df_for_species_chart,'s_ng50','mn:Q','mx:Q','order','yellowgreen',4000,\",.0f\",60,'Scaffold NG50','(Mb)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c2f6b8cb-af1c-4451-bd5f-cfd6612b734c",
   "metadata": {},
   "outputs": [],
   "source": [
    "c_ng50 = hmt(df_for_species_chart,'c_ng50','mn:Q','mx:Q','order','yellowgreen',4000,\",.0f\",60,'Contig NG50','(Mb)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ef7e56dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "sGap = hmt(df_for_species_chart,'sGap','mn:Q','mx:Q','order','yellowgreen',4000,\",.0f\",60, 'Gaps', '(Mb)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "907a29a7-c402-486e-a1e4-11e7da949664",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = alt.hconcat( (class_chart + class_text ),( order_chart + order_text ),( species_chart + species_text))\n",
    "stats = alt.hconcat( size, het, rep,c_ng50, s_ng50, sGap )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0d23af61-b00b-4215-85c8-c890b564a98b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-0dcba55f81494455b61e5b9690272229\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-0dcba55f81494455b61e5b9690272229\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-0dcba55f81494455b61e5b9690272229\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"concat\": {\"spacing\": 0}, \"title\": {\"fontSize\": 9, \"subtitleFontSize\": 8}}, \"hconcat\": [{\"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 1, \"stroke\": \"black\", \"strokeWidth\": 0.5}, \"encoding\": {\"color\": {\"field\": \"class_\", \"legend\": null, \"scale\": {\"domain\": [\"Chondrichthyes\", \"Actinopteri\", \"Amphibia\", \"Reptilia\", \"Lepidosauria\", \"Aves\", \"Mammalia\"], \"range\": [\"#d73027\", \"#fc8d59\", \"#fee090\", \"#ffffbf\", \"#e0f3f8\", \"#91bfdb\", \"#4575b4\"]}, \"type\": \"nominal\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Class\", \"width\": 100}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 12, \"fontWeight\": \"bold\"}, \"encoding\": {\"color\": {\"value\": \"black\"}, \"href\": {\"field\": \"link\", \"type\": \"nominal\"}, \"text\": {\"field\": \"class_\", \"type\": \"nominal\"}, \"y\": {\"field\": \"mid\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Class\", \"transform\": [{\"calculate\": \"('https://www.google.com/search?q=' + datum.class_)\", \"as\": \"link\"}], \"width\": 100}], \"data\": {\"name\": \"data-b718130dd25c796f92203ad038ae0a6a\"}}, {\"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 1, \"stroke\": \"black\", \"strokeWidth\": 0.5}, \"encoding\": {\"color\": {\"field\": \"class_\", \"legend\": null, \"scale\": {\"domain\": [\"Chondrichthyes\", \"Actinopteri\", \"Amphibia\", \"Reptilia\", \"Lepidosauria\", \"Aves\", \"Mammalia\"], \"range\": [\"#d73027\", \"#fc8d59\", \"#fee090\", \"#ffffbf\", \"#e0f3f8\", \"#91bfdb\", \"#4575b4\"]}, \"type\": \"nominal\"}, \"opacity\": {\"field\": \"order\", \"legend\": null, \"type\": \"nominal\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Order\", \"width\": 100}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 10}, \"encoding\": {\"color\": {\"value\": \"black\"}, \"href\": {\"field\": \"link\", \"type\": \"nominal\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"order\", \"type\": \"nominal\"}, \"y\": {\"field\": \"mid\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Order\", \"transform\": [{\"calculate\": \"('https://www.google.com/search?q=' + datum.order)\", \"as\": \"link\"}], \"width\": 100}], \"data\": {\"name\": \"data-6fca83ff158e93152da01a2fca35d624\"}}, {\"layer\": [{\"mark\": {\"type\": \"rect\", \"opacity\": 1, \"stroke\": \"black\", \"strokeWidth\": 0.5}, \"encoding\": {\"color\": {\"field\": \"class_\", \"legend\": null, \"scale\": {\"domain\": [\"Chondrichthyes\", \"Actinopteri\", \"Amphibia\", \"Reptilia\", \"Lepidosauria\", \"Aves\", \"Mammalia\"], \"range\": [\"#d73027\", \"#fc8d59\", \"#fee090\", \"#ffffbf\", \"#e0f3f8\", \"#91bfdb\", \"#4575b4\"]}, \"type\": \"nominal\"}, \"opacity\": {\"field\": \"order\", \"legend\": null, \"type\": \"nominal\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Species\", \"width\": 350}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8, \"fontStyle\": \"italic\"}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"black\", \"test\": \"(datum['class_'] === 'Mammalia')\"}, \"value\": \"black\"}, \"href\": {\"field\": \"link\", \"type\": \"nominal\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"species\", \"type\": \"nominal\"}, \"y\": {\"field\": \"mid\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": \"Species\", \"transform\": [{\"calculate\": \"('https://www.google.com/search?q=' + datum.species)\", \"as\": \"link\"}], \"width\": 350}], \"data\": {\"name\": \"data-5dbbef3d55fbdd2ddde24d88785eb9b8\"}}, {\"hconcat\": [{\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"size\", \"legend\": null, \"scale\": {\"scheme\": \"goldred\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Size\", \"subtitle\": \"(Mb)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['size'] > 4000)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"size\", \"format\": \",.0f\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Size\", \"subtitle\": \"(Mb)\"}, \"width\": 60}]}, {\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"het\", \"legend\": null, \"scale\": {\"scheme\": \"yellowgreen\"}, \"type\": \"nominal\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Het\", \"subtitle\": \"(%)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['het'] > 0.5)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"het\", \"format\": \",.2f\", \"type\": \"nominal\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Het\", \"subtitle\": \"(%)\"}, \"width\": 60}]}, {\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"rep\", \"legend\": null, \"scale\": {\"scheme\": \"yellowgreen\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Repeat\", \"subtitle\": \"(%)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['rep'] > 50)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"rep\", \"format\": \",.1f\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Repeat\", \"subtitle\": \"(%)\"}, \"width\": 60}]}, {\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"c_ng50\", \"legend\": null, \"scale\": {\"scheme\": \"yellowgreen\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Contig NG50\", \"subtitle\": \"(Mb)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['c_ng50'] > 4000)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"c_ng50\", \"format\": \",.0f\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Contig NG50\", \"subtitle\": \"(Mb)\"}, \"width\": 60}]}, {\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"s_ng50\", \"legend\": null, \"scale\": {\"scheme\": \"yellowgreen\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Scaffold NG50\", \"subtitle\": \"(Mb)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['s_ng50'] > 4000)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"s_ng50\", \"format\": \",.0f\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Scaffold NG50\", \"subtitle\": \"(Mb)\"}, \"width\": 60}]}, {\"layer\": [{\"mark\": \"rect\", \"encoding\": {\"color\": {\"field\": \"sGap\", \"legend\": null, \"scale\": {\"scheme\": \"yellowgreen\"}, \"type\": \"quantitative\"}, \"y\": {\"axis\": null, \"field\": \"mn\", \"scale\": {\"domain\": [0, 72], \"nice\": false}, \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Gaps\", \"subtitle\": \"(Mb)\"}, \"width\": 60}, {\"mark\": {\"type\": \"text\", \"align\": \"center\", \"baseline\": \"middle\", \"color\": \"black\", \"dy\": -5, \"fontSize\": 8}, \"encoding\": {\"color\": {\"condition\": {\"value\": \"white\", \"test\": \"(datum['sGap'] > 4000)\"}, \"value\": \"black\"}, \"opacity\": {\"value\": 1}, \"text\": {\"field\": \"sGap\", \"format\": \",.0f\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"mn\", \"type\": \"quantitative\"}, \"y2\": {\"field\": \"mx\"}}, \"height\": 900, \"title\": {\"text\": \"Gaps\", \"subtitle\": \"(Mb)\"}, \"width\": 60}]}], \"data\": {\"name\": \"data-5dbbef3d55fbdd2ddde24d88785eb9b8\"}}], \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-b718130dd25c796f92203ad038ae0a6a\": [{\"class_\": \"Actinopteri\", \"mn\": 4, \"mx\": 16, \"mid\": 9.5}, {\"class_\": \"Amphibia\", \"mn\": 16, \"mx\": 22, \"mid\": 18.5}, {\"class_\": \"Aves\", \"mn\": 33, \"mx\": 53, \"mid\": 42.5}, {\"class_\": \"Chondrichthyes\", \"mn\": 0, \"mx\": 4, \"mid\": 1.5}, {\"class_\": \"Lepidosauria\", \"mn\": 27, \"mx\": 33, \"mid\": 29.5}, {\"class_\": \"Mammalia\", \"mn\": 53, \"mx\": 73, \"mid\": 62.5}, {\"class_\": \"Reptilia\", \"mn\": 22, \"mx\": 27, \"mid\": 24.0}], \"data-6fca83ff158e93152da01a2fca35d624\": [{\"order\": \"Anura\", \"mn\": 16, \"mx\": 22, \"mid\": 18.5, \"class_\": \"Amphibia\"}, {\"order\": \"Artiodactyla\", \"mn\": 53, \"mx\": 60, \"mid\": 56.0, \"class_\": \"Mammalia\"}, {\"order\": \"Atheriniformes\", \"mn\": 4, \"mx\": 5, \"mid\": 4.0, \"class_\": \"Actinopteri\"}, {\"order\": \"Caprimulgiformes\", \"mn\": 33, \"mx\": 34, \"mid\": 33.0, \"class_\": \"Aves\"}, {\"order\": \"Carnivora\", \"mn\": 60, \"mx\": 61, \"mid\": 60.0, \"class_\": \"Mammalia\"}, {\"order\": \"Characiformes\", \"mn\": 5, \"mx\": 9, \"mid\": 6.5, \"class_\": \"Actinopteri\"}, {\"order\": \"Cingulata\", \"mn\": 61, \"mx\": 63, \"mid\": 61.5, \"class_\": \"Mammalia\"}, {\"order\": \"Coliiformes\", \"mn\": 34, \"mx\": 36, \"mid\": 34.5, \"class_\": \"Aves\"}, {\"order\": \"Crocodylia\", \"mn\": 22, \"mx\": 24, \"mid\": 22.5, \"class_\": \"Reptilia\"}, {\"order\": \"Dermoptera\", \"mn\": 63, \"mx\": 64, \"mid\": 63.0, \"class_\": \"Mammalia\"}, {\"order\": \"Didelphimorphia\", \"mn\": 64, \"mx\": 65, \"mid\": 64.0, \"class_\": \"Mammalia\"}, {\"order\": \"Diprotodontia\", \"mn\": 65, \"mx\": 66, \"mid\": 65.0, \"class_\": \"Mammalia\"}, {\"order\": \"Eulipotyphla\", \"mn\": 66, \"mx\": 67, \"mid\": 66.0, \"class_\": \"Mammalia\"}, {\"order\": \"Gruiformes\", \"mn\": 36, \"mx\": 37, \"mid\": 36.0, \"class_\": \"Aves\"}, {\"order\": \"Lampriformes\", \"mn\": 9, \"mx\": 11, \"mid\": 9.5, \"class_\": \"Actinopteri\"}, {\"order\": \"Myliobatiformes\", \"mn\": 0, \"mx\": 4, \"mid\": 1.5, \"class_\": \"Chondrichthyes\"}, {\"order\": \"Opisthocomiformes\", \"mn\": 37, \"mx\": 39, \"mid\": 37.5, \"class_\": \"Aves\"}, {\"order\": \"Passeriformes\", \"mn\": 39, \"mx\": 49, \"mid\": 43.5, \"class_\": \"Aves\"}, {\"order\": \"Perciformes\", \"mn\": 11, \"mx\": 12, \"mid\": 11.0, \"class_\": \"Actinopteri\"}, {\"order\": \"Pholidota\", \"mn\": 67, \"mx\": 69, \"mid\": 67.5, \"class_\": \"Mammalia\"}, {\"order\": \"Proboscidea\", \"mn\": 69, \"mx\": 71, \"mid\": 69.5, \"class_\": \"Mammalia\"}, {\"order\": \"Psittaciformes\", \"mn\": 49, \"mx\": 51, \"mid\": 49.5, \"class_\": \"Aves\"}, {\"order\": \"Rheiformes\", \"mn\": 51, \"mx\": 52, \"mid\": 51.0, \"class_\": \"Aves\"}, {\"order\": \"Scombriformes\", \"mn\": 12, \"mx\": 13, \"mid\": 12.0, \"class_\": \"Actinopteri\"}, {\"order\": \"Siluriformes\", \"mn\": 13, \"mx\": 16, \"mid\": 14.0, \"class_\": \"Actinopteri\"}, {\"order\": \"Sirenia\", \"mn\": 71, \"mx\": 73, \"mid\": 71.5, \"class_\": \"Mammalia\"}, {\"order\": \"Sphenisciformes\", \"mn\": 52, \"mx\": 53, \"mid\": 52.0, \"class_\": \"Aves\"}, {\"order\": \"Squamata\", \"mn\": 27, \"mx\": 33, \"mid\": 29.5, \"class_\": \"Lepidosauria\"}, {\"order\": \"Testudines\", \"mn\": 24, \"mx\": 27, \"mid\": 25.0, \"class_\": \"Reptilia\"}], \"data-5dbbef3d55fbdd2ddde24d88785eb9b8\": [{\"NCBI Species\": \"Acridotheres tristis\", \"Assembly version\": \"hap1\", \"mn\": 39, \"mx\": 40, \"mid\": 39.0, \"species\": \"Acridotheres tristis\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1354.313347, \"het\": 0.558, \"rep\": 30.0, \"s_ng50\": 29.697385, \"c_ng50\": 6.53036, \"sGap\": 3.114772}, {\"NCBI Species\": \"Ammodramus caudacutus\", \"Assembly version\": \"hap1\", \"mn\": 40, \"mx\": 41, \"mid\": 40.0, \"species\": \"Ammodramus caudacutus\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1225.461744, \"het\": 0.236, \"rep\": 21.4, \"s_ng50\": 39.973966, \"c_ng50\": 8.43839, \"sGap\": 4.089548}, {\"NCBI Species\": \"Ammodramus nelsoni\", \"Assembly version\": \"hap1\", \"mn\": 41, \"mx\": 42, \"mid\": 41.0, \"species\": \"Ammodramus nelsoni\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1190.976425, \"het\": 0.571, \"rep\": 21.3, \"s_ng50\": 40.15902, \"c_ng50\": 12.036358, \"sGap\": 5.142283}, {\"NCBI Species\": \"Ara ararauna\", \"Assembly version\": \"hap1\", \"mn\": 49, \"mx\": 50, \"mid\": 49.0, \"species\": \"Ara ararauna\", \"class_\": \"Aves\", \"order\": \"Psittaciformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Ara ararauna\", \"Assembly version\": \"hap2\", \"mn\": 50, \"mx\": 51, \"mid\": 50.0, \"species\": \"Ara ararauna\", \"class_\": \"Aves\", \"order\": \"Psittaciformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Balaenoptera edeni\", \"Assembly version\": \"hap2\", \"mn\": 53, \"mx\": 54, \"mid\": 53.0, \"species\": \"Balaenoptera edeni\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2700.742385, \"het\": 0.0628, \"rep\": 30.3, \"s_ng50\": 108.195663, \"c_ng50\": 35.550969, \"sGap\": 14.987903}, {\"NCBI Species\": \"Candoia aspera\", \"Assembly version\": \"hap1\", \"mn\": 27, \"mx\": 28, \"mid\": 27.0, \"species\": \"Candoia aspera\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1472.456021, \"het\": 0.491, \"rep\": 31.2, \"s_ng50\": 209.78444, \"c_ng50\": 51.781456, \"sGap\": 0.0198}, {\"NCBI Species\": \"Candoia aspera\", \"Assembly version\": \"hap2\", \"mn\": 28, \"mx\": 29, \"mid\": 28.0, \"species\": \"Candoia aspera\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1472.456021, \"het\": 0.491, \"rep\": 31.2, \"s_ng50\": 209.252319, \"c_ng50\": 57.569195, \"sGap\": 0.0188}, {\"NCBI Species\": \"Colius striatus\", \"Assembly version\": \"hap1\", \"mn\": 34, \"mx\": 35, \"mid\": 34.0, \"species\": \"Colius striatus\", \"class_\": \"Aves\", \"order\": \"Coliiformes\", \"size\": 1130.197506, \"het\": 0.646, \"rep\": 12.6, \"s_ng50\": 2.465505, \"c_ng50\": 0.0, \"sGap\": 46.183434}, {\"NCBI Species\": \"Colius striatus\", \"Assembly version\": \"hap2\", \"mn\": 35, \"mx\": 36, \"mid\": 35.0, \"species\": \"Colius striatus\", \"class_\": \"Aves\", \"order\": \"Coliiformes\", \"size\": 1130.197506, \"het\": 0.646, \"rep\": 12.6, \"s_ng50\": 5.445007, \"c_ng50\": 2.47792, \"sGap\": 5.486305}, {\"NCBI Species\": \"Corvus hawaiiensis\", \"Assembly version\": \"hap1\", \"mn\": 42, \"mx\": 43, \"mid\": 42.0, \"species\": \"Corvus hawaiiensis\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1081.864808, \"het\": 0.34399, \"rep\": 10.9934, \"s_ng50\": 68.800448, \"c_ng50\": 23.113903, \"sGap\": 17.498309}, {\"NCBI Species\": \"Cynocephalus volans\", \"Assembly version\": \"hap1\", \"mn\": 63, \"mx\": 64, \"mid\": 63.0, \"species\": \"Cynocephalus volans\", \"class_\": \"Mammalia\", \"order\": \"Dermoptera\", \"size\": 2715.040762, \"het\": 0.611, \"rep\": 26.4, \"s_ng50\": 124.894351, \"c_ng50\": 85.957123, \"sGap\": 0.017}, {\"NCBI Species\": \"Dasypus novemcinctus\", \"Assembly version\": \"hap1\", \"mn\": 61, \"mx\": 62, \"mid\": 61.0, \"species\": \"Dasypus novemcinctus\", \"class_\": \"Mammalia\", \"order\": \"Cingulata\", \"size\": 3235.269131, \"het\": 0.602, \"rep\": 32.5, \"s_ng50\": 89.685286, \"c_ng50\": 14.5185, \"sGap\": 73.679866}, {\"NCBI Species\": \"Dasypus novemcinctus\", \"Assembly version\": \"hap2\", \"mn\": 62, \"mx\": 63, \"mid\": 62.0, \"species\": \"Dasypus novemcinctus\", \"class_\": \"Mammalia\", \"order\": \"Cingulata\", \"size\": 3235.269131, \"het\": 0.602, \"rep\": 32.5, \"s_ng50\": 115.106639, \"c_ng50\": 15.877795, \"sGap\": 30.434832}, {\"NCBI Species\": \"Dixiphia pipra\", \"Assembly version\": \"hap1\", \"mn\": 43, \"mx\": 44, \"mid\": 43.0, \"species\": \"Dixiphia pipra\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 989.883326, \"het\": 1.16, \"rep\": 6.8, \"s_ng50\": 75.758576, \"c_ng50\": 14.46418, \"sGap\": 0.0534}, {\"NCBI Species\": \"Dixiphia pipra\", \"Assembly version\": \"hap2\", \"mn\": 44, \"mx\": 45, \"mid\": 44.0, \"species\": \"Dixiphia pipra\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 989.883326, \"het\": 1.16, \"rep\": 6.8, \"s_ng50\": 75.60774, \"c_ng50\": 11.166331, \"sGap\": 0.0614}, {\"NCBI Species\": \"Dugong dugon\", \"Assembly version\": \"hap1\", \"mn\": 71, \"mx\": 72, \"mid\": 71.0, \"species\": \"Dugong dugon\", \"class_\": \"Mammalia\", \"order\": \"Sirenia\", \"size\": 3158.605109, \"het\": 0.17, \"rep\": 34.8, \"s_ng50\": 177.379183, \"c_ng50\": 57.632671, \"sGap\": 0.0158}, {\"NCBI Species\": \"Dugong dugon\", \"Assembly version\": \"hap2\", \"mn\": 72, \"mx\": 73, \"mid\": 72.0, \"species\": \"Dugong dugon\", \"class_\": \"Mammalia\", \"order\": \"Sirenia\", \"size\": 3158.605109, \"het\": 0.17, \"rep\": 34.8, \"s_ng50\": 138.031769, \"c_ng50\": 57.883746, \"sGap\": 0.0176}, {\"NCBI Species\": \"Emys orbicularis\", \"Assembly version\": \"hap1\", \"mn\": 24, \"mx\": 25, \"mid\": 24.0, \"species\": \"Emys orbicularis\", \"class_\": \"Reptilia\", \"order\": \"Testudines\", \"size\": 2271.648625, \"het\": 0.279, \"rep\": 27.6, \"s_ng50\": 137.743343, \"c_ng50\": 91.255289, \"sGap\": 16.997162}, {\"NCBI Species\": \"Erythrolamprus reginae\", \"Assembly version\": \"hap2\", \"mn\": 29, \"mx\": 30, \"mid\": 29.0, \"species\": \"Erythrolamprus reginae\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1908.066061, \"het\": \"3.74,0.422\", \"rep\": 38.4, \"s_ng50\": 255.449344, \"c_ng50\": 3.202651, \"sGap\": 0.268}, {\"NCBI Species\": \"Eschrichtius robustus\", \"Assembly version\": \"hap1\", \"mn\": 54, \"mx\": 55, \"mid\": 54.0, \"species\": \"Eschrichtius robustus\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2729.218268, \"het\": 0.3008, \"rep\": 32.4, \"s_ng50\": 108.759401, \"c_ng50\": 47.844248, \"sGap\": 0.0124}, {\"NCBI Species\": \"Eubalaena glacialis\", \"Assembly version\": \"hap1\", \"mn\": 55, \"mx\": 56, \"mid\": 55.0, \"species\": \"Eubalaena glacialis\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2661.810066, \"het\": 0.222, \"rep\": 30.4, \"s_ng50\": 84.363911, \"c_ng50\": 33.139823, \"sGap\": 49.090235}, {\"NCBI Species\": \"Eubalaena glacialis\", \"Assembly version\": \"hap2\", \"mn\": 56, \"mx\": 57, \"mid\": 56.0, \"species\": \"Eubalaena glacialis\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2661.810066, \"het\": 0.222, \"rep\": 30.4, \"s_ng50\": 86.797429, \"c_ng50\": 39.169611, \"sGap\": 4.807717}, {\"NCBI Species\": \"Euleptes europaea\", \"Assembly version\": \"hap2\", \"mn\": 30, \"mx\": 31, \"mid\": 30.0, \"species\": \"Euleptes europaea\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1589.842322, \"het\": 0.916, \"rep\": 27.1, \"s_ng50\": 109.232959, \"c_ng50\": 33.587463, \"sGap\": 6.250745}, {\"NCBI Species\": \"Gastrophryne carolinensis\", \"Assembly version\": \"hap1\", \"mn\": 16, \"mx\": 17, \"mid\": 16.0, \"species\": \"Gastrophryne carolinensis\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 4095.803536, \"het\": 1.45, \"rep\": 62.0, \"s_ng50\": 513.356448, \"c_ng50\": 5.281947, \"sGap\": 0.3246}, {\"NCBI Species\": \"Gastrophryne carolinensis\", \"Assembly version\": \"hap2\", \"mn\": 17, \"mx\": 18, \"mid\": 17.0, \"species\": \"Gastrophryne carolinensis\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 4095.803536, \"het\": 1.45, \"rep\": 62.0, \"s_ng50\": 517.518214, \"c_ng50\": 5.065, \"sGap\": 0.3278}, {\"NCBI Species\": \"Gavialis gangeticus\", \"Assembly version\": \"hap1\", \"mn\": 22, \"mx\": 23, \"mid\": 22.0, \"species\": \"Gavialis gangeticus\", \"class_\": \"Reptilia\", \"order\": \"Crocodylia\", \"size\": 2301.125167, \"het\": 0.115, \"rep\": 22.8, \"s_ng50\": 255.848931, \"c_ng50\": 73.46118, \"sGap\": 1.123892}, {\"NCBI Species\": \"Gavialis gangeticus\", \"Assembly version\": \"hap2\", \"mn\": 23, \"mx\": 24, \"mid\": 23.0, \"species\": \"Gavialis gangeticus\", \"class_\": \"Reptilia\", \"order\": \"Crocodylia\", \"size\": 2301.125167, \"het\": 0.115, \"rep\": 22.8, \"s_ng50\": 298.315767, \"c_ng50\": 84.23512, \"sGap\": 1.257318}, {\"NCBI Species\": \"Hippopotamus amphibius\", \"Assembly version\": \"hap1\", \"mn\": 57, \"mx\": 58, \"mid\": 57.0, \"species\": \"Hippopotamus amphibius\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2552.797209, \"het\": 0.001, \"rep\": 23.0, \"s_ng50\": 184.162729, \"c_ng50\": 77.761737, \"sGap\": 0.0166}, {\"NCBI Species\": \"Hippopotamus amphibius\", \"Assembly version\": \"hap2\", \"mn\": 58, \"mx\": 59, \"mid\": 58.0, \"species\": \"Hippopotamus amphibius\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 2552.797209, \"het\": 0.001, \"rep\": 23.0, \"s_ng50\": 168.269428, \"c_ng50\": 72.137584, \"sGap\": 0.0168}, {\"NCBI Species\": \"Hoplias malabaricus\", \"Assembly version\": \"hap1\", \"mn\": 5, \"mx\": 6, \"mid\": 5.0, \"species\": \"Hoplias malabaricus\", \"class_\": \"Actinopteri\", \"order\": \"Characiformes\", \"size\": 1006.60632, \"het\": 0.668, \"rep\": 33.1, \"s_ng50\": 56.848262, \"c_ng50\": 4.154888, \"sGap\": 90.162902}, {\"NCBI Species\": \"Hoplias malabaricus\", \"Assembly version\": \"hap2\", \"mn\": 6, \"mx\": 7, \"mid\": 6.0, \"species\": \"Hoplias malabaricus\", \"class_\": \"Actinopteri\", \"order\": \"Characiformes\", \"size\": 1006.60632, \"het\": 0.668, \"rep\": 33.1, \"s_ng50\": 57.282699, \"c_ng50\": 4.69629, \"sGap\": 61.028905}, {\"NCBI Species\": \"Hyla sarda\", \"Assembly version\": \"hap1\", \"mn\": 18, \"mx\": 19, \"mid\": 18.0, \"species\": \"Hyla sarda\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 3749.586442, \"het\": 1.08, \"rep\": 61.2, \"s_ng50\": 505.300156, \"c_ng50\": 4.426236, \"sGap\": 0.4422}, {\"NCBI Species\": \"Hyla sarda\", \"Assembly version\": \"hap2\", \"mn\": 19, \"mx\": 20, \"mid\": 19.0, \"species\": \"Hyla sarda\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 3749.586442, \"het\": 1.08, \"rep\": 61.2, \"s_ng50\": 419.944051, \"c_ng50\": 4.51617, \"sGap\": 0.4188}, {\"NCBI Species\": \"Hypanus sabinus\", \"Assembly version\": \"hap1\", \"mn\": 0, \"mx\": 1, \"mid\": 0.0, \"species\": \"Hypanus sabinus\", \"class_\": \"Chondrichthyes\", \"order\": \"Myliobatiformes\", \"size\": 3472.076753, \"het\": 0.441, \"rep\": 43.5, \"s_ng50\": 153.447554, \"c_ng50\": 5.354043, \"sGap\": 0.3148}, {\"NCBI Species\": \"Hypanus sabinus\", \"Assembly version\": \"hap2\", \"mn\": 1, \"mx\": 2, \"mid\": 1.0, \"species\": \"Hypanus sabinus\", \"class_\": \"Chondrichthyes\", \"order\": \"Myliobatiformes\", \"size\": 3472.076753, \"het\": 0.441, \"rep\": 43.5, \"s_ng50\": 158.616492, \"c_ng50\": 5.177429, \"sGap\": 0.314}, {\"NCBI Species\": \"Lampris incognitus\", \"Assembly version\": \"hap1\", \"mn\": 9, \"mx\": 10, \"mid\": 9.0, \"species\": \"Lampris incognitus\", \"class_\": \"Actinopteri\", \"order\": \"Lampriformes\", \"size\": 1326.736238, \"het\": 1.06, \"rep\": 50.8, \"s_ng50\": 67.430497, \"c_ng50\": 1.003085, \"sGap\": 0.6238}, {\"NCBI Species\": \"Lampris incognitus\", \"Assembly version\": \"hap2\", \"mn\": 10, \"mx\": 11, \"mid\": 10.0, \"species\": \"Lampris incognitus\", \"class_\": \"Actinopteri\", \"order\": \"Lampriformes\", \"size\": 1326.736238, \"het\": 1.06, \"rep\": 50.8, \"s_ng50\": 61.460152, \"c_ng50\": 1.086565, \"sGap\": 0.6098}, {\"NCBI Species\": \"Liasis olivaceus\", \"Assembly version\": \"hap2\", \"mn\": 31, \"mx\": 32, \"mid\": 31.0, \"species\": \"Liasis olivaceus\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1447.634414, \"het\": 0.186, \"rep\": 17.7, \"s_ng50\": 203.536677, \"c_ng50\": 74.672673, \"sGap\": 5.064312}, {\"NCBI Species\": \"Loxodonta africana\", \"Assembly version\": \"hap1\", \"mn\": 69, \"mx\": 70, \"mid\": 69.0, \"species\": \"Loxodonta africana\", \"class_\": \"Mammalia\", \"order\": \"Proboscidea\", \"size\": 3396.034559, \"het\": 0.32, \"rep\": 26.8, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Loxodonta africana\", \"Assembly version\": \"hap2\", \"mn\": 70, \"mx\": 71, \"mid\": 70.0, \"species\": \"Loxodonta africana\", \"class_\": \"Mammalia\", \"order\": \"Proboscidea\", \"size\": 3396.034559, \"het\": 0.32, \"rep\": 26.8, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Lycodopsis pacificus\", \"Assembly version\": \"hap2\", \"mn\": 11, \"mx\": 12, \"mid\": 11.0, \"species\": \"Lycodopsis pacificus\", \"class_\": \"Actinopteri\", \"order\": \"Perciformes\", \"size\": 590.402093, \"het\": 0.863, \"rep\": 25.8, \"s_ng50\": 28.913003, \"c_ng50\": 5.238746, \"sGap\": 0.0686}, {\"NCBI Species\": \"Macropus eugenii\", \"Assembly version\": \"hap1\", \"mn\": 65, \"mx\": 66, \"mid\": 65.0, \"species\": \"Macropus eugenii\", \"class_\": \"Mammalia\", \"order\": \"Diprotodontia\", \"size\": 3312.89293, \"het\": 0.33294, \"rep\": 25.03798, \"s_ng50\": 488.036699, \"c_ng50\": 16.260088, \"sGap\": 10.151481}, {\"NCBI Species\": \"Manis pentadactyla\", \"Assembly version\": \"hap1\", \"mn\": 67, \"mx\": 68, \"mid\": 67.0, \"species\": \"Manis pentadactyla\", \"class_\": \"Mammalia\", \"order\": \"Pholidota\", \"size\": 2728.549055, \"het\": 0.212, \"rep\": 32.2, \"s_ng50\": 132.345885, \"c_ng50\": 28.762677, \"sGap\": 31.461981}, {\"NCBI Species\": \"Manis pentadactyla\", \"Assembly version\": \"hap2\", \"mn\": 68, \"mx\": 69, \"mid\": 68.0, \"species\": \"Manis pentadactyla\", \"class_\": \"Mammalia\", \"order\": \"Pholidota\", \"size\": 2728.549055, \"het\": 0.212, \"rep\": 32.2, \"s_ng50\": 139.388172, \"c_ng50\": 31.370214, \"sGap\": 14.873332}, {\"NCBI Species\": \"Melospiza georgiana\", \"Assembly version\": \"hap1\", \"mn\": 45, \"mx\": 46, \"mid\": 45.0, \"species\": \"Melospiza georgiana\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1334.824158, \"het\": 1.17, \"rep\": 30.4, \"s_ng50\": 32.012489, \"c_ng50\": 7.607103, \"sGap\": 13.526837}, {\"NCBI Species\": \"Mesoplodon densirostris\", \"Assembly version\": \"hap1\", \"mn\": 59, \"mx\": 60, \"mid\": 59.0, \"species\": \"Mesoplodon densirostris\", \"class_\": \"Mammalia\", \"order\": \"Artiodactyla\", \"size\": 3125.258145, \"het\": 0.293, \"rep\": 42.4, \"s_ng50\": 78.67239, \"c_ng50\": 45.032744, \"sGap\": 0.693296}, {\"NCBI Species\": \"Mobula japanica\", \"Assembly version\": \"hap1\", \"mn\": 2, \"mx\": 3, \"mid\": 2.0, \"species\": \"Mobula japanica\", \"class_\": \"Chondrichthyes\", \"order\": \"Myliobatiformes\", \"size\": 3415.08342, \"het\": 0.358, \"rep\": 43.7, \"s_ng50\": 74.03082, \"c_ng50\": 11.133045, \"sGap\": 35.398234}, {\"NCBI Species\": \"Mobula japanica\", \"Assembly version\": \"hap2\", \"mn\": 3, \"mx\": 4, \"mid\": 3.0, \"species\": \"Mobula japanica\", \"class_\": \"Chondrichthyes\", \"order\": \"Myliobatiformes\", \"size\": 3415.08342, \"het\": 0.358, \"rep\": 43.7, \"s_ng50\": 73.868846, \"c_ng50\": 9.776081, \"sGap\": 66.535008}, {\"NCBI Species\": \"Monodelphis domestica\", \"Assembly version\": \"hap1\", \"mn\": 64, \"mx\": 65, \"mid\": 64.0, \"species\": \"Monodelphis domestica\", \"class_\": \"Mammalia\", \"order\": \"Didelphimorphia\", \"size\": 3216.94597, \"het\": 0.808, \"rep\": 30.8, \"s_ng50\": 518.588076, \"c_ng50\": 4.53615, \"sGap\": 33.385555}, {\"NCBI Species\": \"Neoarius graeffei\", \"Assembly version\": \"hap1\", \"mn\": 13, \"mx\": 14, \"mid\": 13.0, \"species\": \"Neoarius graeffei\", \"class_\": \"Actinopteri\", \"order\": \"Siluriformes\", \"size\": 2420.249501, \"het\": 0.489, \"rep\": 66.6, \"s_ng50\": 80.184264, \"c_ng50\": 14.669962, \"sGap\": 24.860656}, {\"NCBI Species\": \"Neofelis nebulosa\", \"Assembly version\": \"hap1\", \"mn\": 60, \"mx\": 61, \"mid\": 60.0, \"species\": \"Neofelis nebulosa\", \"class_\": \"Mammalia\", \"order\": \"Carnivora\", \"size\": 2395.992815, \"het\": 0.356, \"rep\": 19.9, \"s_ng50\": 149.006193, \"c_ng50\": 54.293935, \"sGap\": 1.357852}, {\"NCBI Species\": \"Odontesthes bonariensis\", \"Assembly version\": \"hap1\", \"mn\": 4, \"mx\": 5, \"mid\": 4.0, \"species\": \"Odontesthes bonariensis\", \"class_\": \"Actinopteri\", \"order\": \"Atheriniformes\", \"size\": 924.220623, \"het\": 0.433, \"rep\": 35.7, \"s_ng50\": 31.128597, \"c_ng50\": 9.61037, \"sGap\": 0.0675}, {\"NCBI Species\": \"Opisthocomus hoazin\", \"Assembly version\": \"hap1\", \"mn\": 37, \"mx\": 38, \"mid\": 37.0, \"species\": \"Opisthocomus hoazin\", \"class_\": \"Aves\", \"order\": \"Opisthocomiformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Opisthocomus hoazin\", \"Assembly version\": \"hap2\", \"mn\": 38, \"mx\": 39, \"mid\": 38.0, \"species\": \"Opisthocomus hoazin\", \"class_\": \"Aves\", \"order\": \"Opisthocomiformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Podargus strigoides\", \"Assembly version\": \"hap1\", \"mn\": 33, \"mx\": 34, \"mid\": 33.0, \"species\": \"Podargus strigoides\", \"class_\": \"Aves\", \"order\": \"Caprimulgiformes\", \"size\": 1092.727389, \"het\": 1.33, \"rep\": 12.0, \"s_ng50\": 85.230406, \"c_ng50\": 14.555328, \"sGap\": 2.220446}, {\"NCBI Species\": \"Poecile atricapillus\", \"Assembly version\": \"hap1\", \"mn\": 46, \"mx\": 47, \"mid\": 46.0, \"species\": \"Poecile atricapillus\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1049.274508, \"het\": 0.893, \"rep\": 10.5, \"s_ng50\": 76.055826, \"c_ng50\": 9.624478, \"sGap\": 1.403275}, {\"NCBI Species\": \"Poecile atricapillus\", \"Assembly version\": \"hap2\", \"mn\": 47, \"mx\": 48, \"mid\": 47.0, \"species\": \"Poecile atricapillus\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1049.274508, \"het\": 0.893, \"rep\": 10.5, \"s_ng50\": 116.828732, \"c_ng50\": 8.200554, \"sGap\": 14.000808}, {\"NCBI Species\": \"Pseudophryne corroboree\", \"Assembly version\": \"hap2\", \"mn\": 20, \"mx\": 21, \"mid\": 20.0, \"species\": \"Pseudophryne corroboree\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 8455.43838, \"het\": 0.471, \"rep\": 73.8, \"s_ng50\": 792.631323, \"c_ng50\": 7.166398, \"sGap\": 22.80747}, {\"NCBI Species\": \"Rhea pennata\", \"Assembly version\": \"hap1\", \"mn\": 51, \"mx\": 52, \"mid\": 51.0, \"species\": \"Rhea pennata\", \"class_\": \"Aves\", \"order\": \"Rheiformes\", \"size\": 1255.251577, \"het\": 0.445, \"rep\": 11.5, \"s_ng50\": 79.541306, \"c_ng50\": 19.335416, \"sGap\": 1.447586}, {\"NCBI Species\": \"Rhynochetos jubatus\", \"Assembly version\": \"hap1\", \"mn\": 36, \"mx\": 37, \"mid\": 36.0, \"species\": \"Rhynochetos jubatus\", \"class_\": \"Aves\", \"order\": \"Gruiformes\", \"size\": 1312.755001, \"het\": 0.602, \"rep\": 22.5, \"s_ng50\": 34.93808, \"c_ng50\": 11.475096, \"sGap\": 0.617785}, {\"NCBI Species\": \"Salminus brasiliensis\", \"Assembly version\": \"hap1\", \"mn\": 7, \"mx\": 8, \"mid\": 7.0, \"species\": \"Salminus brasiliensis\", \"class_\": \"Actinopteri\", \"order\": \"Characiformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Salminus brasiliensis\", \"Assembly version\": \"hap2\", \"mn\": 8, \"mx\": 9, \"mid\": 8.0, \"species\": \"Salminus brasiliensis\", \"class_\": \"Actinopteri\", \"order\": \"Characiformes\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Scomber japonicus\", \"Assembly version\": \"hap1\", \"mn\": 12, \"mx\": 13, \"mid\": 12.0, \"species\": \"Scomber japonicus\", \"class_\": \"Actinopteri\", \"order\": \"Scombriformes\", \"size\": 688.600335, \"het\": 1.22, \"rep\": 19.9, \"s_ng50\": 25.30709, \"c_ng50\": 6.643518, \"sGap\": 0.587}, {\"NCBI Species\": \"Sorex araneus\", \"Assembly version\": \"hap1\", \"mn\": 66, \"mx\": 67, \"mid\": 66.0, \"species\": \"Sorex araneus\", \"class_\": \"Mammalia\", \"order\": \"Eulipotyphla\", \"size\": 2230.886386, \"het\": 1.13646, \"rep\": 27.0295, \"s_ng50\": 338.227856, \"c_ng50\": 4.32518, \"sGap\": 58.832604}, {\"NCBI Species\": \"Spea bombifrons\", \"Assembly version\": \"hap1\", \"mn\": 21, \"mx\": 22, \"mid\": 21.0, \"species\": \"Spea bombifrons\", \"class_\": \"Amphibia\", \"order\": \"Anura\", \"size\": 959.239875, \"het\": 1.73, \"rep\": 28.6, \"s_ng50\": 109.389205, \"c_ng50\": 21.170366, \"sGap\": 6.238534}, {\"NCBI Species\": \"Spheniscus humboldti\", \"Assembly version\": \"hap1\", \"mn\": 52, \"mx\": 53, \"mid\": 52.0, \"species\": \"Spheniscus humboldti\", \"class_\": \"Aves\", \"order\": \"Sphenisciformes\", \"size\": 1246.817591, \"het\": 0.426, \"rep\": 12.6, \"s_ng50\": 88.4601, \"c_ng50\": 15.775369, \"sGap\": 1.417505}, {\"NCBI Species\": \"Sternotherus odoratus\", \"Assembly version\": \"hap1\", \"mn\": 25, \"mx\": 26, \"mid\": 25.0, \"species\": \"Sternotherus odoratus\", \"class_\": \"Reptilia\", \"order\": \"Testudines\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Sternotherus odoratus\", \"Assembly version\": \"hap2\", \"mn\": 26, \"mx\": 27, \"mid\": 26.0, \"species\": \"Sternotherus odoratus\", \"class_\": \"Reptilia\", \"order\": \"Testudines\", \"size\": null, \"het\": null, \"rep\": null, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Taeniopygia guttata\", \"Assembly version\": \"hap1\", \"mn\": 48, \"mx\": 49, \"mid\": 48.0, \"species\": \"Taeniopygia guttata\", \"class_\": \"Aves\", \"order\": \"Passeriformes\", \"size\": 1009.754198, \"het\": 1.47441, \"rep\": 10.8265, \"s_ng50\": 71.125647, \"c_ng50\": 10.156899, \"sGap\": 4.677748}, {\"NCBI Species\": \"Trichomycterus rosablanca\", \"Assembly version\": \"hap1\", \"mn\": 14, \"mx\": 15, \"mid\": 14.0, \"species\": \"Trichomycterus rosablanca\", \"class_\": \"Actinopteri\", \"order\": \"Siluriformes\", \"size\": 998.7785, \"het\": 0.126, \"rep\": 34.1, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Trichomycterus rosablanca\", \"Assembly version\": \"hap2\", \"mn\": 15, \"mx\": 16, \"mid\": 15.0, \"species\": \"Trichomycterus rosablanca\", \"class_\": \"Actinopteri\", \"order\": \"Siluriformes\", \"size\": 998.7785, \"het\": 0.126, \"rep\": 34.1, \"s_ng50\": null, \"c_ng50\": null, \"sGap\": null}, {\"NCBI Species\": \"Vipera latastei\", \"Assembly version\": \"hap1\", \"mn\": 32, \"mx\": 33, \"mid\": 32.0, \"species\": \"Vipera latastei\", \"class_\": \"Lepidosauria\", \"order\": \"Squamata\", \"size\": 1444.460287, \"het\": 0.521, \"rep\": 35.0, \"s_ng50\": 203.43573, \"c_ng50\": 46.203777, \"sGap\": 39.587619}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.HConcatChart(...)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(names | stats).configure_concat(\n",
    "    spacing=0\n",
    ").configure_title(fontSize=9,subtitleFontSize=8)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
